### PR TITLE
Create empty files in parallel for bucketed table

### PR DESCRIPTION
**Pull request by lnyng**
Buckets of a bucketed table in Hive were initialized as empty files serially,
which is very slow.

Now the empty file creation logic in HiveMetadata is factored out as
HiveEmptyFileCreator, which creates files in parallel in a configurable
way similar to HiveStagingFileCommitter. By default the executor has
maximum 100 threads to create files, and can be configured by variable
hive.max-concurrent-file-creates. Considering a 50ms file creation
latency, this feature significantly improves the performance.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/EmptyFileCreator.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.facebook.presto.hive.metastore.Partition;
+import com.facebook.presto.hive.metastore.Table;
+import com.facebook.presto.spi.ConnectorSession;
+import org.apache.hadoop.fs.Path;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface EmptyFileCreator
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/EmptyFileCreator.java]: Unfortunately the methods were called `createEmptyFiles`. However it is not true. It doesn't create empty files. It creates "zero row" files. That's why we need to know all the details about the file format. As with some formats zero rows files are not physically empty.

Let's call this interface `ZeroRowFilesCreator`

File: presto-hive/src/main/java/com/facebook/presto/hive/EmptyFileCreator.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+{
+    void createFiles(ConnectorSession session, Path path, Table table, Optional<Partition> partition,
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/EmptyFileCreator.java]: This method signature can be quite confusing. Mostly because the infuses two different concepts. It is supposed to create zero row files at the given location, but it takes high level logical objects, such as `Table` and `Partition`.

I would recommend changing the signature to something like

```
void create(
  ConnectorSession session,
  HdfsContext hdfsContext,
  Path destinationDirectory,
  List<String> fileNames,
  StorageFormat format,
  HiveCompressionCodec compressionCodec, 
  Properties schema
)
```

File: presto-hive/src/main/java/com/facebook/presto/hive/ForFileCreate.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+@Retention(RUNTIME)
+@Target({FIELD, PARAMETER, METHOD})
+@Qualifier
+public @interface ForFileCreate
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/ForFileCreate.java]: `ForZeroRowFilesCreator`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 4]
```diff
@@ -69,6 +69,7 @@
     private boolean recursiveDirWalkerEnabled;
 
     private int maxConcurrentFileRenames = 20;
+    private int maxConcurrentFileCreates = 100;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: `maxConcurrentZeroRowsFileCreations`

Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: Please change the property name itself respectively

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 5]
```diff
@@ -163,6 +164,18 @@ public ListeningExecutorService createFileRanemeExecutor(HiveConnectorId hiveCli
                                 hiveClientConfig.getMaxConcurrentFileRenames())));
     }
 
+    @ForFileCreate
+    @Singleton
+    @Provides
+    public ListeningExecutorService createFileCreateExecutor(HiveConnectorId hiveClientId, HiveClientConfig hiveClientConfig)
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java]: `createZeroRowFilesCreatorExecutor`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_CLOSE_ERROR;
+import static com.facebook.presto.hive.HiveWriteUtils.initializeSerializer;
+import static com.facebook.presto.hive.metastore.MetastoreUtil.getHiveSchema;
+import static com.facebook.presto.hive.util.ConfigurationUtils.configureCompression;
+import static com.google.common.util.concurrent.Futures.whenAllSucceed;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.util.Objects.requireNonNull;
+
+public class HiveEmptyFileCreator
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java]: `HiveZeroRowFilesCreator`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.emptyFileCreateExecutor = requireNonNull(emptyFileCreateExecutor, "emptyFileCreateExecutor is null");
+    }
+
+    @Override
+    public void createFiles(ConnectorSession session, Path path, Table table, Optional<Partition> partition, List<String> fileNames, HiveCompressionCodec compressionCodec)
+    {
+        HdfsContext hdfsContext = new HdfsEnvironment.HdfsContext(session, table.getDatabaseName(), table.getTableName());
+        JobConf conf = configureCompression(hdfsEnvironment.getConfiguration(hdfsContext, path), compressionCodec);
+        List<ListenableFuture<?>> commitFutures = new ArrayList<>();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java]: nit: Try to keep the lifetime (scope) of a variable small. Keep it close to the first place where it is used.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        try {
+            getFutureValue(listenableFutureAggregate, PrestoException.class);
+        }
+        catch (RuntimeException e) {
+            listenableFutureAggregate.cancel(true);
+            throw e;
+        }
+    }
+
+    private static void writeEmptyFile(ConnectorSession session, Path target, JobConf conf, Properties properties, String serDe, String outputFormatName)
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java]: `createZeroRowsFile`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        // Some serializers such as Avro set a property in the schema.
+        initializeSerializer(conf, properties, serDe);
+
+        // The code below is not a try with resources because RecordWriter is not Closeable.
+        FileSinkOperator.RecordWriter recordWriter = HiveWriteUtils.createRecordWriter(target, conf, properties, outputFormatName, session);
+        try {
+            recordWriter.close(false);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_WRITER_CLOSE_ERROR, "Error write empty file to Hive", e);
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java]: `Error write zero row file to Hive`

**Comment by arhimondr** [comment]
CC: @rschlussel @shixuan-fan

**Comment by arhimondr** [comment]
@lnyng Testcase failure is unrelated and is already fixed on master. Once you rebase this PR on the latest master the failure should be gone

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java [hunk 10]
```diff
@@ -65,6 +65,7 @@ public void testDefaults()
                 .setWriterSortBufferSize(new DataSize(64, Unit.MEGABYTE))
                 .setForceLocalScheduling(false)
                 .setMaxConcurrentFileRenames(20)
+                .setMaxConcurrentFileCreates(100)
```
Comment (rschlussel) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java]: I'm curious how you chose the default? Why is it 5x more than renames? Is it because threads sit idle longer waiting for the file creation to happen? Also, have you tested whether this creates extra load on the coordinator?

Comment (arhimondr) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java]: > I'm curious how you chose the default. Why is it 5x more than renames? Is it because threads sit idle longer waiting for the file creation to happen?

Effectively the first one is used only when commiting a single bucket. The number of files that have to be renamed scales with the number of partitions in the destination table. So it looks like 20 threads should be enough to be able to perform this operation in a reasonable time.

However when inserting into a table that has empty buckets - number of empty files to be created scales as `number of buckets  * number of partitions`, thus the higher number of threads is required to be able to finish it in a reasonable time.

> Also, have you tested whether this creates extra load on the coordinator?

I would expect the threads to be mostly doing IO. We only spend CPU on materializing 0 rows file, what in theory shouldn't be CPU consuming. However that could be file format specific. I think there few things we can do here:

1. Decrease number of threads to something safe (say, 20? or even 10 or 5). 
2. Optimize empty file creating code. Instead of materializing zero-row file content for every file it can be materialized once, and then written in multiple threads to the destination file system. That will make sure that threads will be mostly waiting.
3. Do the #1 first, and then create a separate PR that addresses #2

@lnyng, what do you feel?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+
+    private static void writeEmptyFile(ConnectorSession session, Path target, JobConf conf, Properties properties, String serDe, String outputFormatName)
+    {
+        // Some serializers such as Avro set a property in the schema.
+        initializeSerializer(conf, properties, serDe);
+
+        // The code below is not a try with resources because RecordWriter is not Closeable.
+        FileSinkOperator.RecordWriter recordWriter = HiveWriteUtils.createRecordWriter(target, conf, properties, outputFormatName, session);
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveEmptyFileCreator.java]: nit: static import

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 12]
```diff
@@ -1871,7 +1825,7 @@ public Void visitVariableReference(VariableReferenceExpression variable, Immutab
                                 hivePartitionResult.getEnforcedConstraint(),
                                 hiveBucketHandle,
                                 hivePartitionResult.getBucketFilter())),
-                                hivePartitionResult.getUnenforcedConstraint()));
+                hivePartitionResult.getUnenforcedConstraint()));
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: nit: unrelated change

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 13]
```diff
@@ -163,6 +164,18 @@ public ListeningExecutorService createFileRanemeExecutor(HiveConnectorId hiveCli
...
+    {
+        return listeningDecorator(
+                new ExecutorServiceAdapter(
+                        new BoundedExecutor(
+                                newCachedThreadPool(daemonThreadsNamed("hive-" + hiveClientId + "-%s")),
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java]: How about naming these threads `hive-create-zero-row-file...`, and correspondingly naming the threads in rename to be `hive-rename...`. You could change the rename one in a separate commit before the main commit.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java (hunk 10 above)
Comment (lnyng) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java]: Thanks for the comment. The number is based on that the current default for empty bucket file creation is ~4k, and @arhimondr suggests that the creation latency for each file could be ~50ms. Setting default to 100 reduces the latency from ~3min to ~2s. If using only 20 threads, the latency would be ~10s or more with higher latency or more bucket files. I think 2s latency is acceptable so I pick 100.

When I was testing a single node hive is used and the creation latency is ~500us so the benefit is not significant. But I also manually adding latency to the file creation logic and this new feature works as expected.

For the extra load I was not considering it yet. What metric would you recommend for measuring that?

Comment (arhimondr) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveClientConfig.java]: > For the extra load I was not considering it yet. What metric would you recommend for measuring that?

I think the concern here is that the creating zero rows files might not be just waiting on IO, but actually actively spending CPU cycles. If that's the case - having 100 threads doing it simultaneously may cause some undesired hiccups on coordinator. 

You can try to create multiple zero-rows files in a loop using different storage formats (e.g.: ORC, PARQUET, RCBINARY, etc.) with a reasonable sized table schema (say ~100 different columns). Then you can measure how much CPU time does it take using the operation system cpu counters (`ThreadMXBean#getCurrentThreadCpuTime`) and compare it to the wall time (`System#nanoTime`). If the fraction of the CPU time is not negligible - i would suggest doing the optimization i explained. But again, it doesn't feel like it has to be in this PR. Thus it is up to you.

**Comment by lnyng** [comment]
I have made changes according to suggestions. But travis seems to be unhappy about the change. The error message shown is quite confusing, showing that two empty rows are different. Any suggestion?

**Comment by shixuan-fan** [comment]
@lnyng The failure on `testVirtualBucket` is not related to your change and I restarted the test. I'll discuss with @jessesleeping on this test once he's back.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveZeroRowFileCreator.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+import static com.google.common.util.concurrent.Futures.whenAllSucceed;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.util.Objects.requireNonNull;
+
+public class HiveZeroRowFileCreator
+        implements ZeroRowFileCreator
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final ListeningExecutorService zeroRowFileCreatorExecutor;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveZeroRowFileCreator.java]: nit: simply `executor`

File: presto-hive/src/main/java/com/facebook/presto/hive/ZeroRowFileCreator.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+import com.facebook.presto.hive.metastore.StorageFormat;
+import com.facebook.presto.spi.ConnectorSession;
+import org.apache.hadoop.fs.Path;
+
+import java.util.List;
+import java.util.Properties;
+
+public interface ZeroRowFileCreator
+{
+    void createFiles(ConnectorSession session, HdfsEnvironment.HdfsContext hdfsContext, Path destinationDirectory,
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/ZeroRowFileCreator.java]: nit: We don't use this style of declaring parameters. We either put all the parameters on the same line, or every parameter on a new line

```
void createFiles(
  ConnectorSession session,
  HdfsContext hdfsContext,
  Path destinationDirectory,
  List<String> fileNames,
  StorageFormat format,
  HiveCompressionCodec compressionCodec, 
  Properties schema);
```

Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/ZeroRowFileCreator.java]: nit: static import `HdfsContext`

**Comment by lnyng** [comment]
Rebase and made changes to address nits above

**Comment by arhimondr** [comment]
Thanks @lnyng !

Lets give @rschlussel and @shixuan-fan a chance to also have a look, and then we can merge it.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 16]
```diff
@@ -1403,17 +1373,31 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
...
+            // replace partitionUpdates before creating the zero-row files so that those files will be cleaned up if we end up rollback
             partitionUpdates = PartitionUpdate.mergePartitionUpdates(Iterables.concat(partitionUpdates, partitionUpdatesForMissingBuckets));
+            HdfsContext hdfsContext = new HdfsContext(session, table.get().getDatabaseName(), table.get().getTableName());
             for (PartitionUpdate partitionUpdate : partitionUpdatesForMissingBuckets) {
                 Optional<Partition> partition = table.get().getPartitionColumns().isEmpty() ? Optional.empty() : Optional.of(buildPartitionObject(session, table.get(), partitionUpdate));
-                createEmptyFile(
+
+                Properties schema;
+                StorageFormat format;
+                if (partition.isPresent()) {
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Have we considered extracting this into a separate private method since it seems very similar to the one in `beginCreateTable()`?

**Comment by shixuan-fan** [comment]
@arhimondr Please be aware that this might break the build of our internal extension of Hive connector.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 16 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: How about having two methods:

`Properties getHiveSchema(Table, Optional<Partition>)` and `StorageFormat getStorageFormat(Table, Optional<Partition>)`  

and use them inline?

e.g.:

```
zeroRowFileCreator.createFiles(
                        session,
                        session,
                        hdfsContext,
                        partitionUpdate.getWritePath(),
                        getTargetFileNames(partitionUpdate.getFileWriteInfos()),
                        getStorageFormat(table, partition),
                        handle.getCompressionCodec(),
                       getHiveSchema(table, partition));
```

**Comment by arhimondr** [comment]
@shixuan-fan 

> Please be aware that this might break the build of our internal extension of Hive connector.

That's a good catch. Let me create a migration PR.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 16 above)
Comment (shixuan-fan) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Sounds reasonable

**Comment by lnyng** [comment]
Created two private methods https://github.com/prestodb/presto/pull/13141/commits/3cf66e7f911b835b106796e781423b6c5210edaa#diff-13aa0bdc2b255255b85a1b0763c2cb6dR1213
The method to get schema is called `getSchema` instead of `getHiveSchema` to avoid name collision with the static imported ones that are overloaded. The IDE seems to be confused by overloading two static import method and a private method.

For the concern of breaking internal extension of Hive connector, I guess that should not be addressed in this PR, right? Let me know if there are other issues.

**Comment by arhimondr** [comment]
> I guess that should not be addressed in this PR, right

No, it should not. I will create a migration PR for the internal connector @shixuan-fan mentioned.

**Comment by arhimondr** [comment]
I'm going to merge it after the build passes

**Comment by arhimondr** [comment]
@lnyng Merged, thanks for the contribution!
